src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,763 @@
+      auto nulls = _evaluate_default_null_logic(left.nulls, right.nulls);
+
+      std::vector<R> values(result_size);
+      if (left.size() == right.size()) {
+        for (auto row_idx = ChunkOffset{0}; row_idx < result_size; ++row_idx) {
+          Functor{}(values[row_idx], left.values[row_idx], right.values[row_idx]);
+        }
+      } else if (left.size() > right.size()) {
+        for (auto row_idx = ChunkOffset{0}; row_idx < result_size; ++row_idx) {
+          Functor{}(values[row_idx], left.values[row_idx], right.values.front());
+        }
+      } else {
+        for (auto row_idx = ChunkOffset{0}; row_idx < result_size; ++row_idx) {
+          Functor{}(values[row_idx], left.values.front(), right.values[row_idx]);
+        }
+      }
+
+      result = ExpressionResult<R>{std::move(values), std::move(nulls)};
+    } else {
+      Fail("BinaryOperation not supported on the requested DataTypes");
+    }
+  });
+
+  return result;
+}
+
+template<typename R, typename Functor>
+ExpressionResult<R> ExpressionEvaluator::evaluate_binary_with_custom_null_logic(const AbstractExpression& left_expression,
+                                                         const AbstractExpression& right_expression) {
+  ExpressionResult<R> result;
+
+  resolve_to_expression_result_views(left_expression, right_expression, [&](const auto &left, const auto &right) {
+    using LeftDataType = typename std::decay_t<decltype(left)>::Type;
+    using RightDataType = typename std::decay_t<decltype(right)>::Type;
+
+    if constexpr (Functor::template supports<R, LeftDataType, RightDataType>::value) {
+      const auto result_row_count = _result_size(left, right);
+
+      std::vector<bool> nulls(result_row_count);
+      std::vector<R> values(result_row_count);
+
+      for (auto row_idx = ChunkOffset{0}; row_idx < result_row_count; ++row_idx) {
+        bool null;
+        Functor{}(values[row_idx], null, left.value(row_idx),
+                  left.null(row_idx), right.value(row_idx),
+                  right.null(row_idx));
+        nulls[row_idx] = null;
+      }
+
+      result = ExpressionResult<R>{std::move(values), std::move(nulls)};
+
+    } else {
+      Fail("BinaryOperation not supported on the requested DataTypes");
+    }
+  });
+
+  return result;
+}
+
+template<typename Functor>
+void ExpressionEvaluator::resolve_to_expression_result_views(const AbstractExpression &left_expression,
+                                                             const AbstractExpression &right_expression,
+                                                             const Functor &fn) {
+  resolve_to_expression_results(left_expression, right_expression, [&](const auto& left_result, const auto& right_result) {
+    resolve_expression_result_to_view(left_result, [&](const auto &left_view) {
+      resolve_expression_result_to_view(right_result, [&](const auto &right_view) {
+        fn(left_view, right_view);
```

Comment:
I am not yet sure what to make out of these multiple nested `resolve` thingies. This might be because I have no idea what an "expression result view" might be and how it compares to an "expression result".